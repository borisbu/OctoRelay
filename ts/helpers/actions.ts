import type { Relay, Task } from "../types/Relay";
import { ownCode } from "./const";

export const toggleRelay = (key: string, relay: Relay) => {
  const command = () =>
    OctoPrint.simpleApiCommand(ownCode, "update", { v: 2, subject: key });
  if (!relay.confirm_off) {
    return command();
  }
  const dialog = $("#octorelay-confirmation-dialog");
  dialog.find(".modal-title").text("Turning " + relay.label_text + " off");
  dialog
    .find("#octorelay-confirmation-text")
    .text("Are you sure you want to turn the " + relay.label_text + " off?");
  dialog
    .find(".btn-cancel")
    .off("click")
    .on("click", () => dialog.modal("hide"));
  dialog
    .find(".btn-confirm")
    .off("click")
    .on("click", () => {
      command();
      dialog.modal("hide");
    });
  dialog.modal("show");
};

export const cancelTask = (key: string, { owner, target }: Task) =>
  OctoPrint.simpleApiCommand(ownCode, "cancelTask", {
    v: 2,
    subject: key,
    owner,
    target,
  });
