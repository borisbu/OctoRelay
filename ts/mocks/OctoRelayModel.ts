export const modelMock = jest.fn();

jest.mock("../model/OctoRelayModel", () => ({ OctoRelayViewModel: modelMock }));
