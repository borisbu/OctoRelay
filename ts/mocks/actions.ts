export const cancelMock = jest.fn();

jest.mock("../helpers/actions", () => ({
  cancelTask: cancelMock,
}));
