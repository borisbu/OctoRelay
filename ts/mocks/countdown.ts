export const disposerMock = jest.fn();
export const countdownMock = jest.fn(() => disposerMock);
export const deadlineMock = jest.fn(() => "sample deadline");

jest.mock("../helpers/countdown", () => ({
  setCountdown: countdownMock,
  formatDeadline: deadlineMock,
}));
