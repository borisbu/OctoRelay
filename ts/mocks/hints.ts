export const clearMock = jest.fn();
export const showMock = jest.fn();

jest.mock("../helpers/hints", () => ({
  clearHints: clearMock,
  showHints: showMock,
}));
