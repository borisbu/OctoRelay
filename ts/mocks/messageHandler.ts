export const handlerMock = jest.fn();

jest.mock("../model/messageHandler", () => ({
  makeMessageHandler: jest.fn(() => handlerMock),
}));
