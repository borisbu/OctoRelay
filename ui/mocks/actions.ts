import { vi } from "vitest";

export const cancelMock = vi.fn();
export const toggleMock = vi.fn();

vi.mock("../helpers/actions", () => ({
  cancelTask: cancelMock,
  toggleRelay: toggleMock,
}));
