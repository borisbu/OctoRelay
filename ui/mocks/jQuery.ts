export const elementMock: Record<
  | "toggle"
  | "html"
  | "off"
  | "on"
  | "find"
  | "text"
  | "modal"
  | "is"
  | "popover"
  | "tooltip",
  jest.Mock
> = {
  toggle: jest.fn(() => elementMock),
  html: jest.fn(() => elementMock),
  off: jest.fn(() => elementMock),
  on: jest.fn(() => elementMock),
  find: jest.fn(() => elementMock),
  text: jest.fn(() => elementMock),
  modal: jest.fn(() => elementMock),
  is: jest.fn(() => elementMock),
  tooltip: jest.fn(() => elementMock),
  popover: jest.fn(() => elementMock),
};

export const jQueryMock = jest.fn((subject: string | (() => void)) => {
  if (typeof subject === "function") {
    return subject();
  }
  return elementMock;
});
