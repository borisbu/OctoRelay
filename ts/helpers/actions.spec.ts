import { elementMock, jQueryMock } from "../mocks/jQuery";
import { cancelTask, toggleRelay } from "./actions";

describe("Actions", () => {
  const apiMock = jest.fn();

  Object.assign(global, {
    OctoPrint: { simpleApiCommand: apiMock },
    $: jQueryMock,
  });

  afterEach(() => {
    apiMock.mockClear();
    elementMock.text.mockClear();
    elementMock.modal.mockClear();
    elementMock.find.mockClear();
    elementMock.on.mockClear();
  });

  describe("toggleRelay() action", () => {
    test("should toggle the relay", () => {
      toggleRelay("r1", {
        relay_pin: 16,
        inverted_output: false,
        relay_state: true,
        label_text: "Nozzle Light",
        active: true,
        icon_html: "<div>&#128161;</div>",
        confirm_off: false,
        upcoming: null,
      });
      expect(apiMock).toHaveBeenCalledWith("octorelay", "update", {
        pin: "r1",
      });
    });

    test("should open the confirmation dialog", () => {
      toggleRelay("r2", {
        relay_pin: 12,
        inverted_output: true,
        relay_state: true,
        label_text: "Printer",
        active: true,
        icon_html: '<img src="plugin/dashboard/static/img/printer-icon.png">',
        confirm_off: true,
        upcoming: null,
      });
      expect(jQueryMock).toHaveBeenCalledWith("#octorelay-confirmation-dialog");
      expect(elementMock.find.mock.calls).toMatchSnapshot(".find()");
      expect(elementMock.text.mock.calls).toMatchSnapshot(".text()");
      expect(elementMock.modal).toHaveBeenCalledWith("show");
      expect(elementMock.on).toHaveBeenCalledTimes(2);
      elementMock.modal.mockClear();
      elementMock.on.mock.calls[0][1](); // cancel
      expect(elementMock.modal).toHaveBeenCalledWith("hide");
      elementMock.modal.mockClear();
      elementMock.on.mock.calls[1][1](); // confirm
      expect(apiMock).toHaveBeenCalledWith("octorelay", "update", {
        pin: "r2",
      });
      expect(elementMock.modal).toHaveBeenCalledWith("hide");
    });
  });

  describe("cancelTask() action", () => {
    test("should trigger API command", () => {
      cancelTask("r2", {
        owner: "PRINTING_STARTED",
        target: true,
        deadline: 86400,
      });
      expect(apiMock).toHaveBeenCalledWith("octorelay", "cancelTask", {
        owner: "PRINTING_STARTED",
        subject: "r2",
        target: true,
      });
    });
  });
});
