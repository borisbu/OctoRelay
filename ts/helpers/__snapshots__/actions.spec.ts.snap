// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Actions toggleRelay() action should open the confirmation dialog: .find() 1`] = `
[
  [
    ".modal-title",
  ],
  [
    "#octorelay-confirmation-text",
  ],
  [
    ".btn-cancel",
  ],
  [
    ".btn-confirm",
  ],
]
`;

exports[`Actions toggleRelay() action should open the confirmation dialog: .text() 1`] = `
[
  [
    "Turning Printer off",
  ],
  [
    "Are you sure you want to turn the Printer off?",
  ],
]
`;
